mesonbuild/modules/qt.py

```diff
@@ -71,19 +71,44 @@ def _detect_tools(self, env, method):
                 mlog.log(' {}:'.format(compiler_name.lower()), mlog.red('NO'))
         self.tools_detected = True
 
-    def parse_qrc(self, state, fname):
-        abspath = os.path.join(state.environment.source_dir, state.subdir, fname)
-        relative_part = os.path.dirname(fname)
+    def parse_qrc(self, state, rcc_file):
+        if type(rcc_file) is str:
+            abspath = os.path.join(state.environment.source_dir, state.subdir, rcc_file)
+            relative_part = os.path.dirname(rcc_file)
+        elif type(rcc_file) is File:
+            abspath = rcc_file.absolute_path(state.environment.source_dir, state.environment.build_dir)
+            relative_part = os.path.dirname(abspath)
+
         try:
             tree = ET.parse(abspath)
             root = tree.getroot()
             result = []
             for child in root[0]:
                 if child.tag != 'file':
-                    mlog.warning("malformed rcc file: ", os.path.join(state.subdir, fname))
+                    mlog.warning("malformed rcc file: ", os.path.join(state.subdir, rcc_file))
                     break
                 else:
-                    result.append(os.path.join(relative_part, child.text))
+                    resource_path = child.text
+                    # We need to guess if the pointed resource is:
+                    #   a) in build directory -> implies a generated file
+                    #   b) in source directory
+                    #   c) somewhere else external dependency file to bundle
+                    if os.path.isabs(resource_path):
+                        # a)
+                        if resource_path.startswith(os.path.abspath(state.environment.build_dir)):
+                            resource_relpath = os.path.relpath(resource_path, state.environment.build_dir)
+                            result.append(File(is_built=True, subdir='', fname=resource_relpath))
+                        # either b) or c)
+                        else:
+                            result.append(File(is_built=False, subdir=state.subdir, fname=resource_path))
+                    else:
+                        # a)
+                        if os.path.exists(state.environment.build_dir + resource_path):
```

Comment:
I don't think rcc has a concept of a search path?
Which makes this easier, because for generated files, the files will not exist while meson runs. But i think there is no ambiguity because when a qrc files refers to a file it needs to explicitly use a path to that file. So it should be correct to always look at a absolute version of the path to decide?